### PR TITLE
270 Add courses to CoreSpotlight

### LINKED ISSUE #270: Support App Search
first step: for course entities

### PR DESCRIPTION
**Pull request by jprberlin**
All local courses are searchable by title and description
fixes #270
screen shot 2017-07-27 at 14 40 47

### CODE REVIEW

File: xikolo-common/Data/Helper/SyncHelper.swift [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+    func startObserving() {
+        NotificationCenter.default.addObserver(self, selector: #selector(coreDataChange(note:)), name: NSNotification.Name.NSManagedObjectContextObjectsDidChange, object: CoreDataHelper.viewContext)
+         NotificationCenter.default.addObserver(self, selector: #selector(coreDataChange(note:)), name: NSNotification.Name.NSManagedObjectContextObjectsDidChange, object: CoreDataHelper.backgroundContext)
+        //print("Start oberserving CoreData")
+    }
+
+    func stopObserving() {
+        NotificationCenter.default.removeObserver(self, name: NSNotification.Name.NSManagedObjectContextDidSave, object: CoreDataHelper.viewContext)
+        NotificationCenter.default.removeObserver(self, name: NSNotification.Name.NSManagedObjectContextDidSave, object: CoreDataHelper.backgroundContext)
+        //print("Stop observing CoreData")
```
Comment (bjrne) [inline xikolo-common/Data/Helper/SyncHelper.swift]: Either make this debug, log or remove commented code

File: xikolo-ios/AppDelegate.swift [hunk 2]
```diff
@@ -43,27 +44,65 @@ class AppDelegate : AbstractAppDelegate {
...
+                break
         }
-        
+
         // we can't handle the url, open it with a browser
         let webpageUrl = url
         application.openURL(webpageUrl)
         return false
+
+
```
Comment (bjrne) [inline xikolo-ios/AppDelegate.swift]: Please remove the whitespace

File: xikolo-ios/AppDelegate.swift [hunk 3]
```diff
@@ -43,27 +44,65 @@ class AppDelegate : AbstractAppDelegate {
...
+                        course = try CourseHelper.getBySlug(slug)
+                    } catch {
+                        course = nil
+                    }
+                    if (course != nil){
+                        self.goToCourse(course!)
+                        return true
+                    }
+
+                }else{
```
Comment (bjrne) [inline xikolo-ios/AppDelegate.swift]: I'd prefer our current code style with spaces around braces

File: xikolo-ios/AppDelegate.swift [hunk 4]
```diff
@@ -43,27 +44,65 @@ class AppDelegate : AbstractAppDelegate {
...
+                if url.pathComponents.count > 2{
+                    //support /courses/slug -> course detail page or learning
+                    let slug = url.pathComponents[2]
+                    //get course by slug
+                    //todo the course might not be synced yet, than we could try to fetch from the API by slug
+                    var course: Course?
+                    do {
+                        course = try CourseHelper.getBySlug(slug)
+                    } catch {
+                        course = nil
```
Comment (bjrne) [inline xikolo-ios/AppDelegate.swift]: I think this construct of 5 lines can be abbreviated to
`course = try? CourseHelper.getBySlug(slug)`, but verify that

File: xikolo-common/Data/Helper/SyncHelper.swift (hunk 1 above)
Comment (jprberlin) [inline xikolo-common/Data/Helper/SyncHelper.swift]: removed

File: xikolo-ios/AppDelegate.swift (hunk 4 above)
Comment (jprberlin) [inline xikolo-ios/AppDelegate.swift]: thx

File: xikolo-ios/AppDelegate.swift (hunk 2 above)
Comment (jprberlin) [inline xikolo-ios/AppDelegate.swift]: ok, maybe we could have some linter as part of Pull Requests?

File: xikolo-common/Data/Helper/SyncHelper.swift [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+import Foundation
+import CoreData
+
+
+class SyncHelper {
+    static let standard = SyncHelper()
+
+    func startObserving() {
+        NotificationCenter.default.addObserver(self, selector: #selector(coreDataChange(note:)), name: NSNotification.Name.NSManagedObjectContextObjectsDidChange, object: CoreDataHelper.viewContext)
+         NotificationCenter.default.addObserver(self, selector: #selector(coreDataChange(note:)), name: NSNotification.Name.NSManagedObjectContextObjectsDidChange, object: CoreDataHelper.backgroundContext)
```
Comment (jprberlin) [inline xikolo-common/Data/Helper/SyncHelper.swift]: we dont need to observe the background context

File: xikolo-common/Data/Helper/SearchHelper.swift [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+    }
+    static func setNSUserActivity(course:Course){
+        let activity = NSUserActivity(activityType: self.getReverseDomain(appendix: "course.view"))
+        activity.title = "Viewing Course"
+        activity.requiredUserInfoKeys = ["course_id"]
+        activity.webpageURL = URL(string: Brand.BaseURL + "/courses/" + course.slug!)
+        activity.isEligibleForSearch = true
+        activity.isEligibleForHandoff = true
+        if (course.hidden == false){ // internal courses?
+            activity.isEligibleForPublicIndexing = true
```
Comment (jprberlin) [inline xikolo-common/Data/Helper/SearchHelper.swift]: = !course.hidden

File: xikolo-ios/AppDelegate.swift (hunk 2 above)
Comment (bjrne) [inline xikolo-ios/AppDelegate.swift]: I think @mathebox knows some app for travis. But sometimes added whitespace is intended...

File: xikolo-common/Data/Helper/SearchHelper.swift (hunk 6 above)
Comment (bjrne) [inline xikolo-common/Data/Helper/SearchHelper.swift]: ?

Comment (jprberlin) [inline xikolo-common/Data/Helper/SearchHelper.swift]: wnet through this with max and we noted some improvement possibilities

File: xikolo-ios/AppDelegate.swift (hunk 2 above)
Comment (mathebox) [inline xikolo-ios/AppDelegate.swift]: See #274

**Pull request merged**